### DISCUSSION TITLE
Support other S3 compatible storages?

### DISCUSSION THREAD

**Discussion by lazidoca** [forem/forem#15366] [Feature Requests]
Hi, I am not a Ruby dev but I did check the code and I guess at the moment, Forem only supports AWS S3 storage? There are many other S3 compatible storages such DigitalOcean Space, Linode Object Storage... Is there a plan to support other S3 compatible storages? As I want to selfhost Forem, I also want to use cheaper S3 storage like Linode which is 900% cheaper than AWS, it would be nice if Forem supports it.

Thanks

**Comment by jgaskins** [comment]
@lazidoca You're absolutely right that Forem currently only supports S3. We've actually been discussing supporting other storage providers for a while. We haven't acted on it since to my knowledge we haven't seen a whole lot of demand, so we appreciate you expressing your desire for this feature.

The good news is that, since DO Spaces and Linode Object Storage (and, honestly, most object storage providers that aren't Microsoft Azure) support the S3 API, this should be a very simple change. We just need to enable setting the S3 endpoint. This should be a simple one-line change in the application.

However, the implementation is the easy part. We also have to test, document, and add it to [our selfhost config](https://github.com/forem/selfhost/). That will all take a bit of additional care to make sure it's correct. 😄

**Comment by lazidoca** [comment]
@jgaskins Thank you for your feedback. Can you please give me a hint where and what to adjust/update? May be I can help testing with my usecase

**Reply by jgaskins** [reply]
@lazidoca Do you mean in the code or configuration?

**Reply by lazidoca** [reply]
@jgaskins I mean the code as you stated in your comment:

>  This should be a simple one-line change in the application.

**Reply by jgaskins** [reply]
@lazidoca After looking into the `fog` gem (the adapter we use for file uploads), it looks like [this hash](https://github.com/forem/forem/blob/e5cda633912d9cfcb1539962c65e7abc89b40c1a/config/initializers/carrierwave.rb#L26-L31) just needs this line:

```ruby
endpoint: ApplicationConfig["AWS_S3_ENDPOINT"],
```

If `AWS_S3_ENDPOINT` is not provided in your environment config, it'll fall back to AWS as it does currently since we aren't providing the `endpoint` key at all.

Then, in your config, you can set your `AWS_S3_ENDPOINT` environment variable to the endpoint that shows on your DO Spaces config (note: I may have had to include the `https://` there, but it's entirely possible I'm imagining this) and restart your Forem to pick up the config change:

![Screenshot of DigitalOcean Spaces settings page, with the endpoint information highlighted](https://user-images.githubusercontent.com/108205/142878640-13ac4171-f579-43fb-8cd6-8f8eeff3ea76.png)

**Reply by jgaskins** [reply]
@lazidoca Just confirmed it works on DigitalOcean. You'll need a [DO Spaces key](https://cloud.digitalocean.com/account/api/tokens) to make it work.

I posted #15445 which adds the line to the Ruby configuration. You can run that PR on your selfhost instance by executing `sudo foremimg quay.io/forem/forem:pr-15445`, but keep in mind you'll need to change it back to `:latest` after that PR is merged.

Here's the change I had to make to our `/opt/forem/envs/rails.env` config (in addition to the code change mentioned above):

```bash
AWS_ID=YOUR_DO_SPACES_KEY
AWS_SECRET=YOUR_DO_SPACES_SECRET
AWS_BUCKET_NAME=YOUR_SPACE_NAME
AWS_UPLOAD_REGION=nyc3
AWS_S3_ENDPOINT=https://nyc3.digitaloceanspaces.com
```

Some notes:

- Editing this file requires `sudo`: `sudo vi /opt/forem/envs/rails.env` (feel free to use your favorite terminal editor in place of `vi`)
- The `AWS_UPLOAD_REGION` is the DO region where you created your Space (`nyc3`, `sf2`, etc)
- Turns out we did need the `https://` at the beginning of the `AWS_S3_ENDPOINT` and I wasn't imagining it
- It doesn't look like this configuration supports the DO Spaces CDN because it reads from the same place it writes to. It'd be cool to support this so images can load faster for folks further away from the server, as long as it doesn't add significant maintenance burden for us and is somewhat generalizable for other providers that also provide CDNs.

**Reply by en4839** [reply]
Any instructions for Cloudflare R2 ?
